plugins/GSdx/Renderers/HW/GSTextureCache.cpp

```diff
@@ -2108,6 +2013,122 @@ void GSTextureCache::AttachPaletteToSource(Source* s, uint16 pal, bool need_gs_t
+		sok.surf_m_end_block = t->m_end_block;
+		sok.surf_max_valid_z = t->m_valid.z;
+		sok.surf_max_valid_w = t->m_valid.w;
+	}
+	GL_CACHE("TC comp surf write off: searching offset from %s (0x%x => 0x%x) to %s (0x%x => 0x%x).",
+		write_before_target ? "write" : "target",
+		sok.surf_tex0_tbp0, sok.surf_m_end_block,
+		write_before_target ? "target" : "write",
+		sok.req_bp, sok.req_bp_end);
+	const SurfaceOffset so = ComputeSurfaceOffset(sok);
+	return so;
+}
+
+GSTextureCache::SurfaceOffset GSTextureCache::ComputeSurfaceOffset(const SurfaceOffsetKey& sok)
+{
+	if (sok.req_bp > sok.surf_m_end_block || sok.req_bp_end < sok.surf_tex0_tbp0)
+		return { false, GSVector4i(0, 0, 0, 0) };  // No overlap, early return.
+	auto it = m_surface_offset_cache.find(sok);
+	if (it != m_surface_offset_cache.end())
+		return it->second;  // Cache HIT.
+	
+	// Cache MISS.
+	// Sweep search for a valid <x,y> offset.
+	SurfaceOffset so = { false, GSVector4i(0, 0, 0, 0) };
+	const GSLocalMemory::psm_t& psm_s = GSLocalMemory::m_psm[sok.req_psm];
+	const int surf_max_valid_z = static_cast<int>(sok.surf_max_valid_z);
+	const int surf_max_valid_w = static_cast<int>(sok.surf_max_valid_w);
+	GL_CACHE("TC comp surf off: sweep search <x,y> offset such that BP 0x%x (END 0x%x, Valid <z=%d,w=%d>) + OFF <x,y> = 0x%x.",
+		sok.surf_tex0_tbp0, sok.surf_m_end_block, sok.surf_max_valid_z, sok.surf_max_valid_w, sok.req_bp);
+	for (so.offset.x = 0; so.offset.x < surf_max_valid_z; so.offset.x += psm_s.bs.x)
+	{
+		for (so.offset.y = 0; so.offset.y < surf_max_valid_w; so.offset.y += psm_s.bs.y)
+		{
+			const uint32 candidate_bp = psm_s.bn(so.offset.x, so.offset.y, sok.surf_tex0_tbp0, sok.req_bw);
+			// GL_PERF("TC comp surf off: comparing candidate 0x%x with 0x%x (OFF <%d, %d>)", candidate_bp, sok.req_bp, so.offset.x, so.offset.y);
+			if (sok.req_bp == candidate_bp)
+			{
+				// Sweep search HIT: <x,y> offset found.
+				so.is_valid_offset = true;
+				// Sweep search <z,w> offset.
+				bool computed_zw = false;
+				for (so.offset.z = so.offset.x + psm_s.bs.x - 1; so.offset.z <= surf_max_valid_z; so.offset.z += psm_s.bs.x)
+				{
+					for (so.offset.w = so.offset.y + psm_s.bs.y - 1; so.offset.w <= surf_max_valid_w; so.offset.w += psm_s.bs.y)
+					{
+						const uint32 candidate_bp_end = psm_s.bn(so.offset.z, so.offset.w, sok.surf_tex0_tbp0, sok.req_bw);
+						if (sok.req_bp_end == candidate_bp_end)
+						{
+							computed_zw = true;  // Sweep search HIT: <z,w> offset found.
+							++so.offset.z;
+							++so.offset.w;
+							break;
+						}
+					}
+					if (computed_zw)
+						break;
+				}
+				if (!computed_zw)
+				{
+					// Sweep search MISS: assume max possible <z,w> offset is valid.
+					so.offset.z = surf_max_valid_z;
+					so.offset.w = surf_max_valid_w;
+				}
+				break;
+			}
+		}
+		if (so.is_valid_offset)
+			break;
+	}
+	// Clear cache if size too big.
+	while (m_surface_offset_cache.size() + 1 > S_SURFACE_OFFSET_CACHE_MAX_SIZE)
```

Comment:
why a while instead of an if

Reply:
Given that the map is cleared there is no valid reason, will switch to if